pkg/pipelinerunmetrics/metrics.go

```diff
@@ -0,0 +1,152 @@
+	"knative.dev/pkg/logging"
+	"knative.dev/pkg/metrics"
+)
+
+var (
+	sgCount = stats.Float64(chains.PipelineRunSignedName,
+		chains.PipelineRunSignedDesc,
+		stats.UnitDimensionless)
+
+	sgCountView *view.View
+
+	plCount = stats.Float64(chains.PipelineRunUploadedName,
+		chains.PipelineRunUploadedDesc,
+		stats.UnitDimensionless)
+
+	plCountView *view.View
+
+	stCount = stats.Float64(chains.PipelineRunStoredName,
+		chains.PipelineRunStoredDesc,
+		stats.UnitDimensionless)
+
+	stCountView *view.View
+
+	mrCount = stats.Float64(chains.PipelineRunMarkedName,
+		chains.PipelineRunMarkedDesc,
+		stats.UnitDimensionless)
+
+	mrCountView *view.View
+)
+
+// Recorder holds keys for Tekton metrics
+type Recorder struct {
+	initialized bool
+}
+
+// We cannot register the view multiple times, so NewRecorder lazily
+// initializes this singleton and returns the same recorder across any
+// subsequent invocations.
+var (
+	once           sync.Once
+	r              *Recorder
+	errRegistering error
+)
+
+// NewRecorder creates a new metrics recorder instance
+// to log the PipelineRun related metrics
+func NewRecorder(ctx context.Context) (*Recorder, error) {
+	logger := logging.FromContext(ctx)
+	once.Do(func() {
+		r = &Recorder{
+			initialized: true,
+		}
+		errRegistering = viewRegister()
+		if errRegistering != nil {
+			r.initialized = false
+			logger.Errorf("View Register Failed ", r.initialized)
+			return
+		}
+	})
+
+	return r, errRegistering
+}
+
+func viewRegister() error {
+	sgCountView = &view.View{
+		Description: sgCount.Description(),
+		Measure:     sgCount,
+		Aggregation: view.Count(),
+	}
+
+	plCountView = &view.View{
+		Description: plCount.Description(),
+		Measure:     plCount,
+		Aggregation: view.Count(),
+	}
+
+	stCountView = &view.View{
+		Description: stCount.Description(),
+		Measure:     stCount,
+		Aggregation: view.Count(),
+	}
+
+	mrCountView = &view.View{
+		Description: mrCount.Description(),
+		Measure:     mrCount,
+		Aggregation: view.Count(),
+	}
+	return view.Register(
+		sgCountView,
+		plCountView,
+		stCountView,
+		mrCountView,
+	)
+}
+
+func viewUnregister() {
+	view.Unregister(sgCountView, plCountView, stCountView, mrCountView)
+}
+
+func (r *Recorder) RecordCountMetrics(MetricType string, ctx context.Context) {
+	logger := logging.FromContext(ctx)
+	logger.Debugf("Recording payload count metrics for taskrun ", ctx)
+
+	if !r.initialized {
+		logger.Errorf("ignoring the metrics recording as recorder not initialized ")
+	}
+	switch mt := MetricType; mt {
+	case chains.SGCount:
+		r.countMetrics(ctx, sgCount)
+	case chains.PLCount:
+		r.countMetrics(ctx, plCount)
+	case chains.STCount:
+		r.countMetrics(ctx, stCount)
+	case chains.MRCount:
+		r.countMetrics(ctx, mrCount)
+	}
+
+}
+
+func (r *Recorder) countMetrics(ctx context.Context, measure *stats.Float64Measure) {
+	logger := logging.FromContext(ctx)
+	logger.Debugf("Recording metrcis for Pipelineruns", measure.Description())
+	if !r.initialized {
```

Comment:
Since this is checked in `RecordCountMetrics`, this seems unreachable.

Reply:
Done